doc/sips/remote-metadata/README.md

```diff
@@ -0,0 +1,310 @@
+renter uploads data sectors and corresponding metadata changes in the
+same request. Metadata changes are small because they only change
+the affected bytes of the first sector.
+
+If a sector of a file is discarded or the whole file is removed, then
+the corresponding sector of the host becomes free. When a new sector
+is needed on the host, free sectors are used if they are available.
+If a sector of a file is overwritten, a new version is written to new
+(or free) sector, then parity sectors are written to other hosts,
+after that the old sector and old parity sectors become free.
+`Write` and `Discard` records have `revision` field that is used
+to determine which information about the sector is fresh.
+
+Before renewing a contract, the renter completely removes free sectors
+from the contract by sending Delete RPC and rewrites all the metadata.
+Metadata rewriting is needed because indices of sectors are shifted
+when free data sectors are removed. It is also a good time to skip
+unneeded parts of metadata such as writes that were overwritten
+subsequently.
+
+Parity sectors are added to a group of sectors stored on different
+servers and are uploaded to a non-overlapping set of servers. Metadata
+of a parity sector contains corresponding `Write` records of all data
+sectors involved. If the host holding the data sector is down,
+the renter can recover all data and metadata from parity sectors.
+
+A file is identified with unique integer identifier (e.g. counter).
+After a file is removed, its id is never reused.
+
+### Encryption
+
+Thanks to Merkle root and sector IDs, we don't have to authenticate
+data: if it is tampered by the host (or when transferred to renter),
+the sector hash will be different. So we only need to encrypt it to
+keep safe.
+
+A contract has the following types of sectors:
+
+  * metadata
+  * data
+  * parity
+  * free
+
+A free sector does not hold usable data. When it was not free (i.e.
+was used for something) it was encrypted and that data is still there.
+
+Sectors are encrypted with a stream cipher with random access (i.e.
+which can produce bytes for XOR in the arbitrary location of the
+sequence). One contract corresponds to one sequence (i.e. one
+encryption key). The cipher has large space for its pseudorandom
+sequence. Each sector (metadata, data, and parity) is mapped somewhere
+in the sequence. The exact location is determined by hashing metadata
+of the sector. For a metadata sectors metadata is the number of that
+sector. For data and backup sectors normalized versions of their
+records can be used for that purpose.
+
+Metadata is written by appending records to a metadata sector.
+The unused end of a metadata sector cannot be encrypted with the same
+key (i.e. just have the pseudorandom sequence without any XORed data)
+because otherwise, the host can decrypt the subsequent records by
+XORing them with what was in the affected bytes before. To avoid this
```

Comment:
You may want to look into XTS encryption. It uses a "tweakable block cipher" to enable an interesting property: you can overwrite the sector with new encrypted data without compromising security. The only downside is that it's slower.